app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php

```diff
@@ -0,0 +1,280 @@
+                     $this->messageManager->addErrorMessage($e->getMessage());
+                     continue;
+                } catch (\Exception $e) {
+                    // Other exceptions, log and continue
+                    $this->messageManager->addErrorMessage(__('Could not add product to cart: %1', $e->getMessage()));
+                    // Consider logging $e for debugging
+                    continue;
+                }
+            }
+
+            // Save quote and collect totals
+            $quote->collectTotals();
+            $quote->save();
+
+            // Update checkout session
+            $this->checkoutSession->setQuoteId($quote->getId());
+        }
+
+        // Apply coupon code if provided
+        if (!empty($couponCode)) {
+            try {
+                // Ensure coupon is applied in the context of the potentially switched store
+                $quote->setCouponCode($couponCode)->collectTotals()->save();
+
+                // Check if coupon was actually applied
+                if ($quote->getCouponCode() !== $couponCode) {
+                     // Coupon might be invalid or not applicable to the current store/cart contents
+                    $this->messageManager->addErrorMessage(
+                        __('The coupon code "%1" is not valid or cannot be applied.', $couponCode)
+                    );
+                } else {
+                    // Optionally add a success message
+                    // $this->messageManager->addSuccessMessage(__('Coupon code "%1" was applied.', $couponCode));
+                }
+            } catch (\Exception $e) {
+                 // Log the error for debugging
+                $this->messageManager->addErrorMessage(
+                    __('Could not apply coupon code "%1".', $couponCode)
+                );
+            }
+        }
+
+        // Render the checkout page directly (not a redirect)
+        // This ensures the URL parameters remain in the browser address bar
+        return $this->resultPageFactory->create();
+    }
+
+    /**
+     * Parse the products parameter from the URL
+     *
+     * Format: identifier:qty,identifier:qty
+     * (where identifier can be SKU or product ID)
+     *
+     * @param string $productsParam Products parameter string
+     *
+     * @return array<int, array<string, mixed>>
+     */
+    private function _parseProductsParam(string $productsParam): array
+    {
+        $result = [];
+        $productPairs = explode(',', $productsParam);
+
+        foreach ($productPairs as $pair) {
+            $pair = trim($pair); // Trim whitespace from each pair
+            if (empty($pair)) {
+                continue; // Skip empty entries potentially caused by trailing commas
+            }
+            $parts = explode(':', $pair);
+            if (count($parts) === 2) {
+                $identifier = trim($parts[0]);
+                $qty = filter_var($parts[1], FILTER_VALIDATE_INT); // Validate quantity is an integer
```

Comment:
To prevent misuse, please add a maximum limit on the number of products that can be processed in one request.